### PR TITLE
RouteResponse usage refactoring

### PR DESCRIPTION
**Pull request by Udumft**
### Description
This PR replaces usage of `Route` + Route index in favor of it's original `RouteResponse`. It also removes usage of `Route.routeIdentifier`, as this is a redundant property should be [removed too](https://github.com/mapbox/mapbox-directions-swift/pull/562).

### Implementation
Most of the changes is just a replacing of `IndexedRoute` with `RouteResponse`. But it also forced updating some related entities like `CPTrip.userInfo`, removing `RouteProgress.routeIndex or updating `SessionState` with dedicated `routeIdentifier`. There are many public API changes.

### CODE REVIEW

**Comment by S2Ler** [comment]
This PR will clash with #3159

File: Sources/MapboxCoreNavigation/Router.swift
Comment (S2Ler) [inline Sources/MapboxCoreNavigation/Router.swift:18]: @mapbox/navigation-ios  Does anybody know the reason for using type alias here instead of a full featured type like struct?

Comment (S2Ler) [reply Sources/MapboxCoreNavigation/Router.swift:18]: I think it is a good opportunity to refactor this type in a struct unless there is a strong reason not to.

Comment (Udumft) [reply Sources/MapboxCoreNavigation/Router.swift]: Done

File: Sources/MapboxNavigation/CarPlayManagerDelegate.swift
Comment (S2Ler) [inline Sources/MapboxNavigation/CarPlayManagerDelegate.swift:65]: Documentation is outdated.

Comment (S2Ler) [reply Sources/MapboxNavigation/CarPlayManagerDelegate.swift:65]: I have a strange feeling reading `navigationServiceAlong routeResponse`. How is that possible to navigate along response? It might be a consequence of the wrong abstraction or a better naming can be found.

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (S2Ler) [inline Sources/MapboxCoreNavigation/NavigationService.swift:64]: Why this because optional? Seems wrong that navigation service can operation without a route.

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (S2Ler) [inline Sources/MapboxCoreNavigation/NavigationService.swift:223]: #2589 mentions:
> A convenience initializer for MapboxNavigationService can wrap a Route in a RouteResponse, and a convenience initializer for NavigationViewController can wrap a Route in a RouteResponse and MapboxNavigationService.

Shall we add these convenience inits in this PR? Or what's the reason of not including them?

File: Example/CustomViewController.swift
Comment (S2Ler) [inline Example/CustomViewController.swift:19]: Maybe `var indexedUserRouteResponse: IndexedRouteResponse?` can be used?

File: Sources/MapboxCoreNavigation/RouteProgress.swift
Comment (S2Ler) [inline Sources/MapboxCoreNavigation/RouteProgress.swift:354]: `indexedRouteIndex` still appears in `CodingKeys`.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (S2Ler) [inline Sources/MapboxNavigation/NavigationViewController.swift:60]: I think this code was added to support Storyboards. @MaximAlien please review.

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (S2Ler) [inline Sources/MapboxCoreNavigation/NavigationService.swift:68]: nit: sneaky empty spaces :) I would love swift-format integrated so that we will never see such things ever again.

**Comment by S2Ler** [comment]
@1ec5 Your review is required since you created the original ticket.

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (Udumft) [reply Sources/MapboxCoreNavigation/NavigationService.swift:223]: My motivation not to add these was that all of the routes are supposed to be a result of API calling, and thus are always wrapped in `RouteResponse`. If, for some uncommon reason, user is creating a `Route` object differently, he should take care about correctly [wrapping it](https://github.com/mapbox/mapbox-directions-swift/blob/main/Sources/MapboxDirections/RouteResponse.swift#L41), providing other required fields like options and identifier.

File: Example/CustomViewController.swift
Comment (Udumft) [reply Example/CustomViewController.swift]: Good catch. Fixed

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (Udumft) [reply Sources/MapboxCoreNavigation/NavigationService.swift]: That's right, it's a fallout of early implementation version

File: CHANGELOG.md
Comment (Udumft) [reply CHANGELOG.md]: Thanks for the suggestion

**Review by S2Ler — COMMENTED** [review]
I think the following should be done before merge:
- [ ] Verify that Storyboards integration isn't broken cc @MaximAlien 
- [ ] Review from @1ec5

File: Sources/MapboxCoreNavigation/Router.swift
Comment (1ec5) [inline Sources/MapboxCoreNavigation/Router.swift:20]: Remember to add documentation comments about these properties and also the initializer below so that jazzy recognizes them.

File: Sources/MapboxCoreNavigation/Router.swift
Comment (1ec5) [inline Sources/MapboxCoreNavigation/Router.swift:53]: It made sense to say “`along` `route`” in a sentence, but “`along` `routeResponse`” makes less sense. Maybe “`alongRouteAtIndex`, `in` `routeResponse`”, “`routeResponse`, `routeIndex`”, or simply “`indexedRouteResponse`”.

File: Sources/MapboxCoreNavigation/Router.swift
Comment (1ec5) [inline Sources/MapboxCoreNavigation/Router.swift:68]: If it isn’t valid to set `indexedRouteResponse` directly, let’s make it publicly read-only.

File: Sources/MapboxCoreNavigation/Router.swift
Comment (1ec5) [inline Sources/MapboxCoreNavigation/Router.swift:136]: While we’re at it, the first argument doesn’t need a label.

File: Sources/MapboxNavigation/CPTrip.swift
Comment (1ec5) [inline Sources/MapboxNavigation/CPTrip.swift:6]: RouteResponse has an `options` property that should reflect the options in the request. Do we need to keep a separate `routeOptions` argument around, or can we assume it’s always the same options?

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (1ec5) [inline Sources/MapboxCoreNavigation/NavigationService.swift:221]: `routeOptions` isn’t documented, but I wonder if we still need it, because RouteResponse has an `options` property.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [inline Sources/MapboxNavigation/NavigationViewController.swift:372]: Now that NavigationService contains the full `indexedRouteResponse`, as well as the route options, do we need the application to pass in the route response, route index, and route options separately here, or is it sufficient to pass in just the navigation service?

**Review by 1ec5 — APPROVED** [review]
Meant to attach this approval to https://github.com/mapbox/mapbox-navigation-ios/pull/3182#pullrequestreview-713818693:

I think it would be quite beneficial to remove some redundancy, now that RouteResponse encapsulates more information, but this PR is mergeable as is once the few bits of missing documentation are added.

**Review by MaximAlien — COMMENTED** [review]
This PR brings quite a lot of public API changes. We'll need to create another PR in https://github.com/mapbox/mapbox-navigation-ios-examples to address compilation issues.

**Comment by MaximAlien** [comment]
We'll have to also revise documentation after landing current PR: e.g. in https://docs.mapbox.com/ios/beta/navigation `NavigationViewController` and related APIs are mentioned quite often.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (Udumft) [reply Sources/MapboxNavigation/NavigationViewController.swift:55]: These properties were made read-only to [introduce single entry-point method for changing the route](https://github.com/mapbox/mapbox-navigation-ios/pull/3159): `updateRoute(with:routeOptions:)`. Not sure what to do with storyboard scenario...

File: Sources/MapboxNavigation/CPTrip.swift
Comment (Udumft) [reply Sources/MapboxNavigation/CPTrip.swift:6]: It is expected to be same options.
Previously, we limited this `CPTrip.userInfo` to contain exactly `RouteOptions` object. It is not documented anywhere or, exposed publicly (or at least I could not find any public promise made around it). Technically, we could use `routeResponse.options` and cover both `RouteOptions` and `MatchOptions` cases. 
It seems rather safe change to me, so I'll update it if there's no objections.

Comment (Udumft) [reply Sources/MapboxNavigation/CPTrip.swift:6]: Looks like same applies to `waypoints` argument. It is extracted from the route response

File: Tests/CocoaPodsTest/PodInstall/Podfile.lock
Comment (Udumft) [reply Tests/CocoaPodsTest/PodInstall/Podfile.lock:71]: My bad. Will update CocoaPods

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (Udumft) [reply Sources/MapboxCoreNavigation/NavigationService.swift:221]: This argument serves as specification that it handles `RouteOptions` instead of possible `MatchOptions`.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (Udumft) [reply Sources/MapboxNavigation/NavigationViewController.swift:372]: Do you mean a convenience init which would get these values from the custom navigation service provided? Sounds good to me.
But I think we still need to preserve this initializer. Otherwise we'll force users to explicitly instantiate the navigation service even if they are happy with the default one.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (1ec5) [reply Sources/MapboxNavigation/NavigationViewController.swift:55]: #3159 was primarily concerned with changing the route during a trip, but the storyboard code path is about setting up the trip in the first place. Can we introduce something specific to storyboards, perhaps even with “forStoryboard” in the name?

Comment (S2Ler) [reply Sources/MapboxNavigation/NavigationViewController.swift:55]: Would be good to have a special method which setups vc after view controller. This method will set appropriate private variables (could be even grouped in a struct) which will be read in the code which setups vc in storyboards case. 

When I read this code I found it confusing why route variable has such strange get/set.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (MaximAlien) [inline Sources/MapboxNavigation/NavigationViewController.swift:391]:
```suggestion
     This method basically re-runs the setup which takes place in `init`. It could be useful if some of the attributes have changed before `NavigationViewController` did load it's view, or if you did not have access to initializing logic. For example, as a part of `UIStoryboardSegue` configuration.
```

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (MaximAlien) [inline Sources/MapboxNavigation/NavigationViewController.swift:400]: Why do we need to perform check to see if view was really loaded? I recall that in previous implementation in https://github.com/mapbox/mapbox-navigation-ios/pull/2974 it wasn't needed. Just want to make sure that this will not cause any side effects.

File: Tests/MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift
Comment (MaximAlien) [inline Tests/MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift:39]: It'd be great to create a follow-up ticket to update documentation everywhere. In scope of this PR files like `README.md` and `cover.md` can be updated as they show how to use `NavigationViewController`.

File: Sources/MapboxNavigation/NavigationViewController.swift
Comment (Udumft) [reply Sources/MapboxNavigation/NavigationViewController.swift:400]: This check is added since updating `routeResponse`, `routeIndex` or `routeOptions` will not have effect on the contents `NavigationViewController` if view is already loaded. In this case `view` - is the Navigation View, and these properties are only stored in `NavigationViewController` to initialize the NavigationView. So, this check lets user know if this update actually happened or not. Previous implementation did silent update of a local property, but that did not affect the NavigationView if it was already loaded.

File: Tests/MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift
Comment (Udumft) [reply Tests/MapboxCoreNavigationTests/MapboxCoreNavigationTests.swift:39]: I missed that, thanks

File: Sources/MapboxCoreNavigation/RouteProgress.swift
Comment (MaximAlien) [inline Sources/MapboxCoreNavigation/RouteProgress.swift:192]: We need to update documentation here as well. `routeIndex` is no longer present. `options` and `spokenInstructionIndex` will have to be added as well.

**Review by MaximAlien — APPROVED** [review]
This PR changes so many public APIs that it's hard to see what can be wrong. On high level seems to look good to me. We should review all docs and update examples to make sure that there is not major regression and everything is up-to-date.

**Comment by Udumft** [comment]
Ticket to update examples project : mapbox/mapbox-navigation-ios-examples/issues/135

File: Sources/MapboxCoreNavigation/NavigationService.swift
Comment (dmiluski) [reply Sources/MapboxCoreNavigation/NavigationService.swift:224]: Hi @Udumft ,

I have a use case in which I would prefer injecting a Route and am wondering if I now need to create a convenience init to support? It seems strange to move from a strong required Route, to a model of optionals?

https://github.com/mapbox/straightaway-ios/blob/main/Straightaway/Modules/Map/ContinuousNavigationViewController.swift#L85

Given our UI/ViewModels are intentionally separate from network models, exposing a network response object seemed heavy. 

For additional context, in an upcoming feature change, we will be displaying multiple direction options. Which we will display, and begin navigation upon selection of a given route. 



cc: @irtemed88

Comment (dmiluski) [reply Sources/MapboxCoreNavigation/NavigationService.swift:224]: Hi @Udumft ,

I was seeing if I could get your suggestion for how I could approach this? I took a first stab at the upgrade migration.
https://github.com/mapbox/straightaway-ios/pull/1047

**Pull request merged**
